### PR TITLE
add WriteNewVariantsTable task

### PR DESCRIPTION
**Pull request by jklugherz**
- `WriteNewVariantsTable` is a new task which inherits from BaseWriteTask. It identifies new variants from the callset, annotates them with rdc annotations, and writes to a new table.
- `UpdateVariantAnnotationsTableWithNewSamplesTask` now requires `WriteNewVariantsTable` and instead of joining variants from the callset directly it joins from variants from the new variants table. vep still runs here.

The new task is covered in existing tests in `update_variant_annotations_table_with_new_samples_test.py` because I didn't mock the dependency.

### CODE REVIEW

**Review by bpblanken — COMMENTED** [review]
looks good!

File: v03_pipeline/lib/annotations/annotations_utils.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import hail as hl
+
+from hail_scripts.utils.mapping_gene_ids import load_gencode
```
[the remark this answers is not in the document]
Comment (bpblanken) [reply v03_pipeline/lib/annotations/annotations_utils.py:3]: There are two copies of `load_gencode` so the old code remains deployable.

File: v03_pipeline/lib/tasks/base/base_variant_annotations_table.py [hunk 2]
```diff
@@ -16,20 +20,8 @@
...
-        ):
-            annotation_dependencies[f'{rdc.value}_ht'] = hl.read_table(
-                valid_reference_dataset_collection_path(
-                    self.reference_genome,
-                    self.dataset_type,
-                    rdc,
-                ),
-            )
-        return annotation_dependencies
+    def annotation_dependencies(self) -> dict[str, hl.Table]:
```
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/base/base_variant_annotations_table.py]: Yeah, I actually needed to do it again and pushed a change right after you reviewed. `UpdateVariantAnnotationsTableWithUpdatedReferenceDataset` has rdc tables as dependencies but doesn't need the lookup table or gencode mapping. 

I put `other_annotation_dependencies` back into `UpdateVariantAnnotationsTableWithNewSamplesTask` where it has been since the recent change.

File: v03_pipeline/lib/tasks/write_new_variants_table.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        ]
+        if Env.REFERENCE_DATA_AUTO_UPDATE:
+            upstream_table_tasks.append(
+                UpdateVariantAnnotationsTableWithUpdatedReferenceDataset(
+                    self.reference_genome,
+                    self.dataset_type,
+                    self.sample_type,
+                ),
+            )
+        if self.dataset_type.has_lookup_table:
```
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:84]: This is true. I had both here so that `BaseVariantAnnotationsTableTask` would call `initialize_table` to create an empty annotations table in test but there's a better way to do that

File: v03_pipeline/lib/tasks/write_new_variants_table.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    UpdateLookupTableTask,
+)
+from v03_pipeline.lib.tasks.write_remapped_and_subsetted_callset import (
+    WriteRemappedAndSubsettedCallsetTask,
+)
+
+VARIANTS_PER_VEP_PARTITION = 1e3
+
+
+class WriteNewVariantsTable(BaseWriteTask):
```
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:34]: did you mean to suggest something else? it's already `WriteNewVariantsTable`

[the remark this answers is not in the document]
Comment (bpblanken) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:34]: oh yeah, `WriteNewVariantsTableTask` 🤣

File: v03_pipeline/lib/tasks/write_new_variants_table.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bpblanken) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:84]: Does `UpdateVariantAnnotationsTableWithUpdatedReferenceDataset` call `initialize_table` under the hood?

[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:84]: My previous answer was wrong - ` initialize_table` is called but also `update_table` is as well, and in `BaseVariantAnnotationsTableTask` `update_table` simply returns `ht`. If I don't mock out the `UpdateVariantAnnotationsTableWithUpdatedReferenceDataset` dependency in tests it runs its implementation of `update_table` which is a can of worms and doesn't need to happen in those tests 

I have an idea of how to fix it though

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 5]
```diff
@@ -161,41 +107,18 @@ def complete(self) -> bool:
...
+        new_variants_ht = hl.read_table(
+            new_variants_table_path(
+                self.reference_genome,
+                self.dataset_type,
+            ),
         )
-        callset_ht = callset_ht.distinct()
 
-        # 1) Get new rows and annotate with vep
+        # Get new rows and annotate with vep.
```
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:117]: I kept vep in `UpdateVariantAnnotationsTableWithNewSamplesTask` because I want to slot the new clingen registering step in between `WriteNewVariantsTable` and `UpdateVariantAnnotationsTableWithNewSamplesTask` - which means vep would run only if registering succeeds.

[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:117]: I guess VEP really should be in its own task, so maybe now's the time to do that

[the remark this answers is not in the document]
Comment (bpblanken) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:117]: yeah... there's definitely some uncertainty here.  I think at this point we have three separate things that we do for the new variants: 1) clingen register 2) VEP 3) formatting annotations.  We could put all three in one task, or split them apart.

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples_test.py [hunk 6]
```diff
@@ -215,7 +241,10 @@ def test_multiple_update_vat(
         mock_update_rdc_task: Mock,
     ) -> None:
         mock_update_rdc_task.return_value = MockCompleteTask()
-        mock_update_vat_with_rdc_task.return_value = MockCompleteTask()
+        mock_update_vat_with_rdc_task.side_effect = create_blank_annotations_table(
```
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples_test.py]: yeah, this wasn't right. mocking it to `BaseVariantAnnotationsTableTask` works here.

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jklugherz) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py]: Now, all the new variant stuff is in `WriteNewVariantsTableTask`!

[the remark this answers is not in the document]
Comment (bpblanken) [reply v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py]: 🤟

File: v03_pipeline/lib/paths.py [hunk 7]
```diff
@@ -236,3 +236,18 @@ def variant_annotations_table_path(
...
     )
+
+
+def new_variants_table_path(
+    reference_genome: ReferenceGenome,
+    dataset_type: DatasetType,
+    run_id: str,
+) -> str:
+    return os.path.join(
+        runs_path(
```
Comment (jklugherz) [inline v03_pipeline/lib/paths.py:247]: moved the `new_variants.ht` into the `runs` directory

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 8]
```diff
@@ -52,84 +34,25 @@ class UpdateVariantAnnotationsTableWithNewSamplesTask(BaseVariantAnnotationsTabl
...
-                    self.dataset_type,
-                ),
-            )
-
-        if self.dataset_type.has_gencode_mapping:
-            annotation_dependencies['gencode_mapping'] = hl.literal(
-                load_gencode(GENCODE_RELEASE, ''),
-            )
-        return annotation_dependencies
+    run_id = luigi.Parameter()
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:37]: note to self: pass this parameter in airflow

File: v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py [hunk 9]
```diff
@@ -161,92 +84,47 @@ def complete(self) -> bool:
...
+                    lookup_ht=hl.read_table(
+                        lookup_table_path(
+                            self.reference_genome,
+                            self.dataset_type,
+                        ),
+                    ),
                     **self.param_kwargs,
                 ),
             )
+            ht = ht.union(callset_variants_ht, unify=True)
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/update_variant_annotations_table_with_new_samples.py:125]: I had to add `unify=True` to handle the theoretical case if `ht` is empty and all the variants are new, `gt_stats` won't be in a column in the table yet and `union` fails but I think `unify=True` makes sense for this case.

File: v03_pipeline/lib/tasks/write_new_variants_table.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        for rdc in ReferenceDatasetCollection.for_reference_genome_dataset_type(
+            self.reference_genome,
+            self.dataset_type,
+        ):
+            if rdc.requires_annotation:
+                continue
+            rdc_ht = self.annotation_dependencies[f'{rdc.value}_ht']
+            new_variants_ht = new_variants_ht.join(rdc_ht, 'left')
+
+        # 3) TODO: Send to clingen allele registry.
```
Comment (jklugherz) [inline v03_pipeline/lib/tasks/write_new_variants_table.py:201]: I'm going to add this step in a subsequent PR!

File: v03_pipeline/lib/tasks/write_new_variants_table.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+            ),
+        )
+        new_variants_ht = run_vep(
+            new_variants_ht,
+            self.dataset_type,
+            self.reference_genome,
+        )
+
+        # 5) Select down to the formatting annotations fields and any reference dataset collection annotations.
+        rdc_annotations = [
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/write_new_variants_table.py:223]: This block got added because we've flipped the order of the join and the formatting annotations right?  Was there a reason for that?  

I don't think it matters too much... but probably slightly better for VEP to run before the join.  The join tends to be safe and VEP fragile... so running the fragile thing first means earlier failure.

Comment (jklugherz) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:223]: > This block got added because we've flipped the order of the join and the formatting annotations right? 

That's correct, and the order was flipped because I was going to rely on the rdc annotations to register alleles before vep, but now that I'm pretty sure registering will rely on vep, I will return the order to what it was before I created the new task.

Comment (bpblanken) [reply v03_pipeline/lib/tasks/write_new_variants_table.py:223]: 👍 yeah, I think that makes sense!

File: v03_pipeline/lib/tasks/write_metadata_for_run.py [hunk 12]
```diff
@@ -47,7 +47,7 @@ def output(self) -> luigi.Target:
     def complete(self) -> bool:
         return GCSorLocalTarget(self.output().path).exists()
 
-    def requires(self) -> luigi.Task:
+    def requires(self) -> list[luigi.Task]:
```
Comment (bpblanken) [inline v03_pipeline/lib/tasks/write_metadata_for_run.py:50]: 🙇

**Review by bpblanken — APPROVED** [review]

File: v03_pipeline/lib/tasks/write_metadata_for_run.py (hunk 12 above)
Comment (jklugherz) [reply v03_pipeline/lib/tasks/write_metadata_for_run.py:50]: ah I just realized we need it in addition to `ruff`

Comment (bpblanken) [reply v03_pipeline/lib/tasks/write_metadata_for_run.py:50]: yeah, I have some context on why we don't have it right now if you want it, but probably better on a ticket.  

tldr neither hail nor luigi play well with strict type-checking (airflow does though I think).
